public/static/js/legacyMain.js

```diff
@@ -80,14 +80,40 @@ const listenScrollToTopButton = () => {
   });
 };
 
-const detectEnviromentAndSetDownloadOptions = () => {
+const detectEnviromentAndSetDownloadOptions = async () => {
+  /**
+   * This function is used to get the CPU bitness of the user's device
+   * on Windows 11. For more information, see:
+   *
+   * @see
+   * [MDN](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorUAData/getHighEntropyValues)
+   * and
+   * [MSFT](https://learn.microsoft.com/en-us/microsoft-edge/web-platform/how-to-detect-win11)
+   *
+   * @returns
+   * 32 or 64, depending on the CPU bitness of the user's device.
+   * If not supported, it returns undefined.
+   */
+  async function getCPUBitnessForWin11() {
```

Comment:
I think instead of making a complex asynchronous function we can simply:

```js
const userAgent = navigator.userAgent;
const userAgentData = navigator.userAgentData;
const osMatch = userAgent.match(/(Win|Mac|Linux)/);

const os = (osMatch && osMatch[1]) || '';

// detects the architecture through regular ways on user agents that 
// expose the architecture and platform information
// @note this is not available on Windows11 anymore
// @see https://learn.microsoft.com/en-us/microsoft-edge/web-platform/how-to-detect-win11
let arch === userAgent.match(/x86_64|Win64|WOW64/) ? 'x64' : 'x86';

// detects the platform through a legacy property on navigator
// available only on firefox
// @see https://developer.mozilla.org/en-US/docs/Web/API/Navigator/oscpu
if (navigator.oscpu && navigator.oscpu.length) {
  arch = navigator.oscpu.match(/x86_64|Win64|WOW64/) ? 'x64' : 'x86';
}

// detects the platform through a legacy property on navigator
// only available on internet explorer
if (navigator.cpuClass && navigator.cpuClass.length) {
  arch = navigator.cpuClass === 'x64' ? 'x64' : 'x86';
}

// detects the architecture and other platform data on the navigator
// available only on Chromium-based browsers
// @see https://developer.mozilla.org/en-US/docs/Web/API/NavigatorUAData/getHighEntropyValues
if (userAgentData && userAgentData.getHighEntropyValues) {
  // note that getHighEntropyValues returns an object
  const platform = await userAgentData.getHighEntropyValues(['bitness']);

  if (platform && platform.bitness) {
    // note that platform.bitness returns a string
    arch = platform.bitness === "64" ? 'x64' : 'x86';
  }
}

...
```

Reply:
In my snippet above you have a full "safe" and easy-to-read and understand approach. I'd appreciate if you update your PR with this snippet.

Reply:
> In my snippet above you have a full "safe" and easy-to-read and understand approach. I'd appreciate if you update your PR with this snippet.

Sorry but what do you mean by "safe" here? Do you think this method, if an async one, will be unsafe to be called?

Reply:
I don't want to argue here but look at your snippet's complexity versus mine's simplicity.

Reply:
> Do you mean to remove the async function and then paste what you've written directly inside "detectEnviromentAndSetDownloadOptions"?

Nope.

> If so, the function "userAgentData.getHighEntropyValues" is a Promise-based (awaitable one). It must be inside an async method. So "detectEnviromentAndSetDownloadOptions" should still be an async one.

My original comment about not making it asynchronous was based if we only used promises. But then I made a snippet with a nice clean code using async/await. So, I apologise for the confusion.

> If according to your flattern code, because each browser has its own checking way, if mixed together, some checking ways matches while the others not, you'll get an error result in the end. We don't need to go through each "if" and assign to "arch".

Ehh, not really. If you look carefully, you'll notice that:

1. the `oscpu` check only runs on Firefox
2. the `cpuClass` will only run on IE11 and prior
3. The userAgentData will only run on Chromium

Meaning, no condition is being superseded. You can test this manually to reach your own conclusions.

Reply:
Got it, thanks I'll fix mine.